Wait for cancel before next listen in switchLatest

lrhn opened the pull request:
If a stream is performing resource cleanup on cancel it may be a problem
to listen to the next stream and start consuming resources before the
cleanup is done.

lrhn committed:
Wait for cancel before next listen in switchLatest

If a stream is performing resource cleanup on cancel it may be a problem
to listen to the next stream and start consuming resources before the
cleanup is done.

lrhn commented:
This is an alternative version using `pause` on the outer stream while cancelling. The logic is simpler, and all streams are listened to (even if they're then almost immediately cancelled).

natebosch commented on lib/src/switch.dart:
```diff
@@ -61,20 +61,38 @@ extension SwitchLatest<T> on Stream<Stream<T>> {
+        if (currentSubscription != null) {
+          innerSubscription = null;
+          outerSubscription.pause();
+          currentSubscription.cancel().whenComplete(() {
```
Thinking more about it I'm leaning towards forwarding errors to the result stream.

lrhn replied:
That should be possible. I'd do something like:
```dart
Future<void>.sync(currentSubscription.cancel).catchError(controller.addError).whenComplete(() {
  ...
});
```
That should also catch synchronous errors from `cancel` (just to be paranoid).
I *think* it's safe to use `controller.addError` unconditionally. There is no chance that it has been closed if we paused `outerStream` when we received the previous stream event which caused the current cancel.

natebosch committed:
Reformat

natebosch commented on test/switch_test.dart:
```diff
@@ -49,16 +51,12 @@ void main() {
-          first
-            ..add(1)
-            ..add(2);
+          first..add(1)..add(2);
```
It looks like your SDK has an older version of dart format. If you update to the latest dev SDK the formatting should match CI.

natebosch commented:
@googlebot I consent

lrhn committed:
Propagate cancel errors to the result stream.

Also handle cancels on the (non-broadcast) result stream that happen
during inner-stream cancels correctly.

lrhn commented:
PTAL. Added forwarding of errors.

natebosch commented:
I'm working with some teams to resolve some internal failures before we land this.
